src/Dropdown/MultiDropdownWithSelectAll.tsx

```diff
@@ -0,0 +1,162 @@
+import * as React from "react";
+import Select, { components } from "react-select";
+
+import Creatable from "react-select/lib/Creatable";
+import { Props } from "react-select/lib/Creatable";
+import * as cx from "classnames";
+import {
+  CommonProps,
+  defaultProps,
+  defaultComponents,
+  defaultStyle,
+  getCommonClassnames,
+  DefaultProps
+} from "./commons";
+import { GroupType, ValueType } from "react-select/lib/types";
+
+export function allSelected<OptionType = never>(): SelectAllValue<OptionType> {
+  return { type: "AllSelected" };
+}
+
+export function someSelected<OptionType>(values: OptionType[]) {
+  return { type: "SomeSelecetd", values };
+}
+
+type SelectAllOptionType = { value: "_ALL"; label: string };
+type SelectAllValue<OptionType> =
+  | {
+      type: "AllSelected";
+    }
+  | {
+      type: "SomeSelected";
+      values: OptionType[];
+    };
+
+type NonDefaultProps<OptionType> = CommonProps<
+  OptionType | SelectAllOptionType
+> & {
+  selectAllLabel: string;
+  value: SelectAllValue<OptionType>;
+  onChange: (value: SelectAllValue<OptionType>) => void;
+};
+
+export namespace MultiDropdownWithSelectAll {
+  export type Props<OptionType> = NonDefaultProps<OptionType> &
+    Partial<DefaultProps>;
+}
+
+function isGroupedOptionsArray<OptionType>(
+  options: Array<GroupType<OptionType>> | Array<OptionType>
+): options is Array<GroupType<OptionType>> {
+  return options.length > 0 && (options[0] as any).hasOwnProperty("options");
+}
+
+export class MultiDropdownWithSelectAll<OptionType> extends React.PureComponent<
+  NonDefaultProps<OptionType> & DefaultProps
+> {
+  static defaultProps = defaultProps;
+
+  selectAllOption: SelectAllOptionType = {
+    value: "_ALL",
+    label: this.props.selectAllLabel
+  };
+
+  injectSelectAllOptions = (
+    options: MultiDropdownWithSelectAll.Props<OptionType>["options"]
+  ):
+    | Array<OptionType | SelectAllOptionType>
+    | Array<GroupType<OptionType | SelectAllOptionType>> => {
+    if (options === undefined || options.length === 0) return [];
```

Comment:
the `options === undefined` check should probably be removed?

Reply:
`options` is defined as optional in react-select.
I'll overwrite its type in `CommonProps` to be `NonNullable`.